crates/sui-indexer/src/handlers/objects_snapshot_processor.rs

```diff
@@ -62,105 +92,220 @@ impl Default for SnapshotLagConfig {
-                        latest_fn_cp = self.client.get_latest_checkpoint().await?.sequence_number;
-                        start_cp = self
-                            .store
-                            .get_latest_object_snapshot_checkpoint_sequence_number()
-                            .await?
-                            .unwrap_or_default();
-                    }
-                }
-            }
-        }
+    // Receives object changes from the ingestion pipeline and commits them to the store,
+    // keeping the appropriate amount of checkpoint lag behind the rest of the indexer.
+    pub async fn commit_objects_snapshot(
+        store: S,
+        watermark: CheckpointSequenceNumber,
+        indexed_obj_receiver: Receiver<CheckpointObjectChanges>,
+        commit_notifier: watch::Sender<Option<CheckpointSequenceNumber>>,
+        metrics: IndexerMetrics,
+        config: SnapshotLagConfig,
+        cancel: CancellationToken,
+    ) -> IndexerResult<()> {
+        let batch_size = config.snapshot_max_lag - config.snapshot_min_lag;
+        let mut stream = mysten_metrics::metered_channel::ReceiverStream::new(indexed_obj_receiver)
+            .ready_chunks(batch_size);
+
+        let mut start_cp = watermark;
 
-        info!("Objects snapshot processor starts updating objects_snapshot periodically...");
+        info!("Starting objects snapshot committer...");
         loop {
             tokio::select! {
-                _ = self.cancel.cancelled() => {
+                _ = cancel.cancelled() => {
                     info!("Shutdown signal received, terminating object snapshot processor");
                     return Ok(());
                 }
-                _ = tokio::time::sleep(std::time::Duration::from_secs(self.config.sleep_duration)) => {
-                    let latest_cp = self
-                        .store
+                _ = tokio::time::sleep(std::time::Duration::from_secs(1)) => {
+                    let latest_indexer_cp = store
                         .get_latest_checkpoint_sequence_number()
                         .await?
                         .unwrap_or_default();
 
-                    if latest_cp > start_cp + self.config.snapshot_max_lag as u64 {
-                        info!("Objects snapshot processor is updating objects snapshot table from {} to {}", start_cp, start_cp + snapshot_window);
-                        self.store
-                            .update_objects_snapshot(start_cp, start_cp + snapshot_window)
-                            .await?;
-                        start_cp += snapshot_window;
-                        self.metrics
-                            .latest_object_snapshot_sequence_number
-                            .set(start_cp as i64);
+                    // We update the snapshot table when it falls behind the rest of the indexer by more than the max_lag.
+                    while latest_indexer_cp > start_cp + config.snapshot_max_lag as u64 {
```

Comment:
no biggies but to be exact, here should be >= instead, b/c start_cp is already the next cp to process, let's say the latest snapshot is 300, here start_cp is 301 and we can start this once latest_indexer_cp reaches 901?